data/transactions/logic/debugger.go

```diff
@@ -16,294 +16,143 @@
 
 package logic
 
-import (
-	"bytes"
-	"crypto/sha256"
-	"encoding/base64"
-	"encoding/hex"
-	"fmt"
-	"net/http"
-	"net/url"
-	"strings"
-
-	"github.com/algorand/go-algorand/config"
-	"github.com/algorand/go-algorand/data/basics"
-	"github.com/algorand/go-algorand/data/transactions"
-	"github.com/algorand/go-algorand/logging"
-	"github.com/algorand/go-algorand/protocol"
-)
-
-// DebuggerHook functions are called by eval function during TEAL program execution
-// if provided
+// DebuggerHook functions are called by eval function during TEAL program execution, if a debugger
+// is provided.
+//
+// Refer to the lifecycle graph below for the sequence in which hooks are called.
+//
+// NOTE: Arguments given to Debugger hooks (EvalParams and EvalContext) are passed by reference,
+// they are not copies. It is therefore the responsibility of the debugger hooks to NOT modify the
+// state of the structs passed to them. Additionally, hooks are responsible for copying the information
+// they need from the argument structs. No guarantees are made that the referenced state will not
+// change between hook calls. This decision was made in an effort to reduce the performance
+// impact of the debugger hooks.
+//
+//   LOGICSIG LIFECYCLE GRAPH
+//   ┌─────────────────────────┐
+//   │ LogicSig Evaluation     │
+//   ├─────────────────────────┤
+//   │ > BeforeLogicEval       │
+//   │                         │
+//   │  ┌───────────────────┐  │
+//   │  │ Teal Operation    │  │
+//   │  ├───────────────────┤  │
+//   │  │ > BeforeTealOp    │  │
+//   │  │                   │  │
+//   │  │ > AfterTealOp     │  │
+//   │  └───────────────────┘  │
+//   |   ⁞  ⁞  ⁞  ⁞  ⁞  ⁞  ⁞   │
+//   │                         │
+//   │ > AfterLogicEval        │
+//   └─────────────────────────┘
+//
+//   APP LIFECYCLE GRAPH
+//   ┌────────────────────────────────────────────────┐
+//   │ Transaction Evaluation                         │
+//   ├────────────────────────────────────────────────┤
+//   │ > BeforeTxn                                    │
+//   │                                                │
+//   │  ┌──────────────────────────────────────────┐  │
+//   │  │ ? App Call                               │  │
+//   │  ├──────────────────────────────────────────┤  │
+//   │  │ > BeforeLogicEval                        │  │
+//   │  │                                          │  │
+//   │  │  ┌────────────────────────────────────┐  │  │
+//   │  │  │ Teal Operation                     │  │  │
+//   │  │  ├────────────────────────────────────┤  │  │
+//   │  │  │ > BeforeTealOp                     │  │  │
```

Comment:
I'll comment elsewhere about this but we should avoid "Teal" here. Perhaps `BeforeOpcode` works.

Reply:
Changed to `BeforeOpcode` and `AfterOpcode`.

I also renamed `BeforeLogicEval` and `AfterLogicEval` to `BeforeProgram` and `AfterProgram`, to make it clear that they are called for both apps and LogicSigs.